File: frontend/src/types/order.ts
import type { AddressRequest } from "./auth";
import type { CardBrand } from "./payment";

export type OrderStatus = "PREPARING";

export type PaymentMethod = "CARD" | "CASH_ON_DELIVERY";

/** A single order line item (mirrors backend {@code OrderItemResponse}). */
export interface OrderItem {
  id: number;
  productId: number;
  productName: string;
  productImageUrl: string | null;
  unitPrice: number;
  quantity: number;
  lineTotal: number;
}

/** Order aggregate (mirrors backend {@code OrderResponse}). */
export interface Order {
  id: number;
  status: OrderStatus;
  addressId: number;
  subtotal: number;
  discountAmount: number;
  total: number;
  paymentMethod: PaymentMethod;
  cardBrand: CardBrand | null;
  cardLast4: string | null;
  couponCode: string | null;
  createdAt: string;
  items: OrderItem[];
}

/** Request body for {@code POST /api/v1/orders}. */
export interface PlaceOrderRequest {
  addressId?: number;
  newAddress?: AddressRequest;
  paymentMethod: PaymentMethod;
  cardNumber?: string;
  cardholderName?: string;
  expiryMonth?: number;
  expiryYear?: number;
  cvv?: string;
  couponCode?: string;
}

File: backend/src/main/java/com/platform/ecommerce/order/domain/Order.java
package com.platform.ecommerce.order.domain;

import com.platform.ecommerce.user.domain.Address;
import com.platform.ecommerce.user.domain.User;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.CascadeType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;

/**
 * A customer order. Deliberately stores only the card brand and last four
 * digits — never the full card number or CVV — exactly like a real payment
 * processor exposes to merchants.
 */
@Entity
@Table(name = "orders")
@Getter
@Setter
@NoArgsConstructor
public class Order {

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  private Long id;

  @ManyToOne(fetch = FetchType.LAZY, optional = false)
  @JoinColumn(name = "user_id", nullable = false)
  private User user;

  /** The delivery address snapshot (reuses the existing Address entity). */
  @ManyToOne(fetch = FetchType.LAZY, optional = false)
  @JoinColumn(name = "address_id", nullable = false)
  private Address address;

  @Enumerated(EnumType.STRING)
  @Column(nullable = false, length = 20)
  private OrderStatus status = OrderStatus.PREPARING;

  @Column(nullable = false, precision = 12, scale = 2)
  private BigDecimal subtotal;

  @Column(name = "discount_amount", nullable = false, precision = 12, scale = 2)
  private BigDecimal discountAmount = BigDecimal.ZERO;

  @Column(nullable = false, precision = 12, scale = 2)
  private BigDecimal total;

  @Enumerated(EnumType.STRING)
  @Column(name = "payment_method", nullable = false, length = 30)
  private PaymentMethod paymentMethod;

  @Enumerated(EnumType.STRING)
  @Column(name = "card_brand", length = 20)
  private CardBrand cardBrand;

  @Column(name = "card_last4", length = 4)
  private String cardLast4;

  @Column(name = "coupon_code", length = 50)
  private String couponCode;

  @CreationTimestamp
  @Column(name = "created_at", nullable = false, updatable = false)
  private Instant createdAt;

  @OneToMany(mappedBy = "order", fetch = FetchType.LAZY,
      cascade = CascadeType.ALL, orphanRemoval = true)
  private List<OrderItem> items = new ArrayList<>();
}